### DISCUSSION TITLE
Is there a way to render the console output inside the GUI?

### DISCUSSION THREAD

**Discussion by KieranDevvs** [tui-cs/Terminal.Gui#1986] [Q&A] [answered]
Looking for a way to render the console output to the user within the GUI.

**Comment by tznind** [comment, marked as answer]
Do you mean you want a gui.cs `View` which shows the output of a console process? Do you need it to be interactive (i.e. a subwindow that is a fully interactive console) or can it be readonly?

I have in the past redirected the output stream from a seperate (console) readonly `Process` to Terminal.Gui (see https://github.com/HicServices/RDMP/blob/develop/Tools/rdmp/CommandLine/Gui/Windows/RunnerWindows/RunEngineWindow.cs).  This used a ListView and redirected output streams to list items.  But there was no ability for the user to provide interactive feedback (e.g. Console.ReadLine).


If you could explain more about your requirements that would help.

**Reply by KieranDevvs** [reply]
No user input, just purely the output buffer as text that the user can see within the GUI (with scrolling and text selection). You'll have to forgive my lack of knowledge on this library terminology, Ive only started using it today. I think its a `view` that I require.

**Reply by tznind** [reply]
Ok here is a minimum code example of capturing the output of a console app subprocess:

First create a testing app called Echo.exe:

```csharp
for(int i=0;i<100000;i++)
{
    Console.WriteLine("Hello" + i);
    Thread.Sleep(1000);
}
```

This app will compile to Echo.exe and output a series of statements 1s appart.

Compile the app an confirm that it works in a normal console to output as expected.

Next create a Terminal.Gui project and add the following code:

```csharp
using System.Diagnostics;
using Terminal.Gui;

// init Terminal.Gui
Application.Init();

var w = new Window();

// list of data items
List<string> output = new();

// threading safety (optional)
object lockList = new();

// Terminal.Gui View that shows items in the list 'output'
var list = new ListView(output)
{
    Width = Dim.Fill(),
    Height = Dim.Fill()
};

// add the Terminal.Gui list to the main window
w.Add(list);


// Start a subprocess that is the Echo.exe application
var process = new Process
{
    StartInfo = new ProcessStartInfo
    {
        // what to run
        FileName = @"D:\Repos\Echo\Echo\bin\Debug\net6.0\Echo.exe",

        //redirect the output to this application
        UseShellExecute = false,
        RedirectStandardOutput = true,
        CreateNoWindow = true
    }
};

// start the Echo.exe application
process.Start();



Task.Run(() =>
{
    // while the Ehco.exe process has not closed its output stream
    while (!process.StandardOutput.EndOfStream)
    {
        // read a line from the stdout of theprocess
        var line = process.StandardOutput.ReadLine().Trim();

        lock (lockList)
        {
            // append it to the list of outputs we know about
            output.Insert(0, line);
            
            //update the UI
            Application.MainLoop.Invoke(() => list.SetNeedsDisplay());
        }

    }
});

Application.Run(w);

Application.Shutdown();
```

This should show:
![processexe](https://user-images.githubusercontent.com/31306100/188288883-6e59ec63-e0f0-4291-bdb7-d0f7523b803f.gif)

This should be a good starter for outputting what you are looking for?  From here you may decide to instead use `TextView` or write your own custom subclass of `View`. 

Let me know if this example is confusing or if you are unable to make it work

**Reply by KieranDevvs** [reply]
If you use `Console.SetOut()` and define a text writer, you can basically write the current process console output buffer to whatever you want. This avoids having to have a separate process / shell handle.

```cs
Console.SetOut(new TextViewWriter(_myTextView));
```

```cs
public class TextViewWriter : TextWriter
{
    private readonly TextView _textView;

    public TextViewWriter(TextView textView)
    {
        _textView = textView;
    }

    public override void Write(char value)
    {
        WriteText(value.ToString());
    }

    public override void Write(string? value)
    {
        WriteText(value);
    }

    public override void WriteLine(char value)
    {
        WriteText($"{value}{Environment.NewLine}");
    }

    public override void WriteLine(string? value)
    {
        WriteText($"{value}{Environment.NewLine}");
    }

    public override Encoding Encoding => Encoding.Unicode;

    private void WriteText(string? text)
    {
        _textView.InsertText(text);
    }
}
```

The problem im having is that the `TextView` doesn't render the new lines (everything is on the first line).

![image](https://user-images.githubusercontent.com/21192520/188289580-81d642d5-01fb-443e-bcd4-ff2d55bd1f1e.png)

**Reply by tznind** [reply]
> If you use `Console.SetOut()` and define a text writer, you can basically write the current process console output buffer to whatever you want. This avoids having to have a separate process / shell handle.

Oh that is cool, way cleaner.  Thanks for show me that, I will use that in future.

So firstly there seems to be a bug with `InsertText` and newlines (I've opened a ticket here #1987).  You should be able to use the `Text` property directly though (e.g. `textView.Text += "newtext"`.

Secondly I think you should use  `Application.MainLoop.Invoke(() => dosomething);` when updating the UI from your writer.  Invoke ensures the `Action` occurs on the UI thread.  Although the code might work without, I think you will have stability issues later on.

You may run into usability issues with TextView (loosing users selection / scroll location during update).  I'm also not sure how well it will scale to very large volumes/rates of text.  This is why I went with `ListView`.  But that said, there's no reason TextView shouldn't be able to handle this use case.  I did explore something similar to this in 
#1838 and adjust some of the current behaviors that are problematic in this scenario (scrolling text) but haven't had the time to finish it up properly and it still had issues.

**Reply by KieranDevvs** [reply]
Thanks for the help, got it working somewhat. I just need to get scrolling to work and get text colouring to work but these are separate issues that i'll investigate on my own.

![image](https://user-images.githubusercontent.com/21192520/188319128-c08afe46-50d2-45c2-b703-2595ee2ee100.png)

![image](https://user-images.githubusercontent.com/21192520/188319156-39ec7433-a1b9-48a9-b045-f4455a113b43.png)
